www/src/components/email-capture-form.js

```diff
@@ -1,9 +1,182 @@
+      marginBottom: `calc(1.05rem / 2)`,
+      color: colors.warning,
+      fontSize: rhythm(1 / 2),
+    }}
+  >
+    {children}
+  </div>
+)
+
+class Form extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.onSubmit = this.onSubmit.bind(this)
+
+    // let's use uncontrolled components https://reactjs.org/docs/uncontrolled-components.html
+    this.email = React.createRef()
+  }
+
+  state = {
+    errorMessage: ``,
+    fieldErrors: {},
+  }
+
+  onSubmit(e) {
+    e.preventDefault()
+
+    // validation here or use Formik or something like that
+    // we can also rely on server side validation like I do right now
+
+    const { portalId, formId, sfdcCampaignId } = this.props
+    const url = `https://api.hsforms.com/submissions/v3/integration/submit/${portalId}/${formId}`
+    const data = {
+      fields: [
+        {
+          name: `email`,
+          value: this.email.current.value,
+        },
+      ],
+      context: {
+        pageUri: window.location.href,
+        pageName: document.title,
+      },
+    }
+
+    if (sfdcCampaignId) {
+      data.context.sfdcCampaignId = sfdcCampaignId
+    }
+
+    const xhr = new XMLHttpRequest()
+    xhr.open(`POST`, url, false)
+    xhr.setRequestHeader(`Content-Type`, `application/json`)
+
+    xhr.onreadystatechange = () => {
+      if (xhr.readyState == 4) {
+        const response = JSON.parse(xhr.responseText)
+        if (xhr.status == 200) {
+          // https://developers.hubspot.com/docs/methods/forms/submit_form_ajax
+          // docs mention response should "thankYouMessage" field,
+          // but I get inlineMessage in my testing
+          this.props.onSuccess(
+            response.thankYouMessage || response.inlineMessage
+          )
+        } else {
+          let errorMessage,
+            fieldErrors = {}
+          if (response.errors) {
+            // {"message":"Error in 'fields.email'. Invalid email address","errorType":"INVALID_EMAIL"}
+            // {"message":"Error in 'fields.email'. Required field 'email' is missing","errorType":"REQUIRED_FIELD"}
+
+            const errorRe = /^Error in 'fields.([^']+)'. (.+)$/
+            response.errors.map(error => {
+              const [, fieldName, message] = errorRe.exec(error.message)
+              fieldErrors[fieldName] = message
+            })
+
+            // hubspot use "The request is not valid" message, so probably better use custom one
+            errorMessage = `Errors in the form`
+          } else {
+            errorMessage = response.message
+          }
+
+          this.setState({ fieldErrors, errorMessage })
+        }
+      }
+    }
+
+    xhr.send(JSON.stringify(data))
+  }
+
+  render() {
+    return (
+      <form onSubmit={this.onSubmit}>
+        <div
+          css={{
+            paddingBottom: `20px`,
+          }}
+        >
+          <Label isRequired for="email">
+            Email
+          </Label>
+          <SingleLineInput
```

Comment:
and here I forgot to set `id` (so that `for` prop would actually work)

Reply:
👍 🙏  Will fix!